pkg/storagespace/storagespace.go

```diff
@@ -0,0 +1,151 @@
+// Copyright 2018-2021 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package storagespace
+
+import (
+	"path"
+	"strings"
+
+	provider "github.com/cs3org/go-cs3apis/cs3/storage/provider/v1beta1"
+	"github.com/cs3org/reva/v2/pkg/utils"
+	"github.com/pkg/errors"
+)
+
+const (
+	_idDelimiter        string = "!"
+	_storageIDDelimiter string = "$"
+)
+
+var (
+	// ErrInvalidSpaceReference signals that the space reference is invalid.
+	ErrInvalidSpaceReference = errors.New("invalid storage space reference")
+	// ErrInvalidSpaceID signals that the space ID is invalid.
+	ErrInvalidSpaceID = errors.New("invalid storage space id")
+)
+
+// SplitID splits a storage space ID into a provider ID and a node ID.
+// The accepted formats results of the storage space ID and respective results
+// are:
+// <providerid>$<spaceid>!<nodeid> 	-> <providerid>$<spaceid>, <nodeid>
+// <providerid>$<spaceid>			-> <providerid>$<spaceid>, <spaceid>
+// <spaceid>						-> <spaceid>, <spaceid>
+func SplitID(ssid string) (storageid, nodeid string, err error) {
```

Comment:
This func is pretty complicated
1) I need to read the comment to use it
2) It can unwrap `storageid!opaqueid` but that is not obvious

Imho it would be better to have a func that only does `storageid!opaqueid` -> `storageid, opaqueid` It would be much cleaner what this function is supposed to do.

Reply:
The thing is that we have implicitly defined the behavior that if there is no node ID we will take the space ID as the node ID.
Sure we can make this function dumb and just split at "!" but then we have to implement that logic with using the space ID as the node ID multiple times in different places.

Reply:
This behaviour doesn't sound correct to me in all places. If the nodeID is the spaceID then we are querying the space root. But this should not be default behaviour. A client should clarify that it wants to query the space root and not hope that the server repairs the request for him.

Reply:
In our ~~offline~~ off-GitHub discussion we clarified that this is the current implemented behavior. I.e. it already is like this in edge. We have also documented this and therefor will leave it as it is.